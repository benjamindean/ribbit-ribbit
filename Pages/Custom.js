// @flow
import React, { PureComponent } from 'react';
import { ToastAndroid, Vibration, View } from 'react-native';
import { Col, Row, Grid } from 'react-native-easy-grid';
import Stopwatch from 'timer-stopwatch';
import Button from '../Components/Button';
import PlayButton from '../Components/PlayButton';
import PlayProgressBar from '../Components/PlayProgressBar';
import * as PlayProgressBarActions from '../Actions/PlayProgressBar';
import * as PatternActions from '../Actions/Pattern';
import styles from '../Styles/main';

import { Provider, connect } from 'react-redux';
import { createStore, bindActionCreators } from 'redux';
import rootReducer from '../Reducers/index';

class MainView extends PureComponent {
	componentDidMount() {
		this.vibrateTime = new Stopwatch();
		this.pausedTime = new Stopwatch();
	}

	_updatePattern = time => {
		this.props.patternobject.update(time);
	};

	_onPressIn = () => {
		this.pausedTime.stop();
		this._updatePattern(this.pausedTime.ms);
		this.pausedTime.reset();

		this.vibrateTime.start();
		Vibration.vibrate(5000);
	};

	_onPressOut = () => {
		this.pausedTime.start();

		this.vibrateTime.stop();
		this._updatePattern(this.vibrateTime.ms);
		Vibration.cancel();

		this.vibrateTime.reset();
	};

	_resetPattern = () => {
		Vibration.cancel();

		this.props.patternobject.reset();
		this.props.playprogressbar.reset();

		this.pausedTime.stop();
		this.pausedTime.reset();
		this.vibrateTime.stop();
		this.vibrateTime.reset();

		ToastAndroid.show('Pattern cleared', ToastAndroid.SHORT);
	};

	_updateProgressBar = () => {
		this.props.playprogressbar.reset();

		const percent = this.props.pattern.reduce((a, b) => a + b, 0) / 10;
		const interval = setInterval(() => {
			if (this.props.progress < 0.9) {
				this.props.playprogressbar.update(0.1);
			} else {
				clearInterval(interval);
			}
		}, percent);
	};

	_playPattern = () => {
		if (this.props.pattern.length) {
			this._updateProgressBar();

			ToastAndroid.show(
				`Playing pattern: ${JSON.stringify(this.props.pattern)}`,
				ToastAndroid.LONG
			);

			Vibration.vibrate(this.props.pattern);
		}
	};

	render() {
		return (
			<Grid style={styles.container}>
				<Col>
					<View style={styles.statusbar} />
					<Row size={1} />
					<Row size={2}>
						<Col>
							<View style={styles.recordButton}>
								<PlayProgressBar
									progress={this.props.progress}
									pattern={this.props.pattern}
								/>
								<PlayButton
									onPressIn={this._onPressIn}
									onPressOut={this._onPressOut}
								/>
							</View>
						</Col>
					</Row>

					<Row size={2}>
						<Col>
							<Button
								onPress={this._playPattern}
								name="controller-play"
							/>
						</Col>
						<Col>
							<Button onPress={this._resetPattern} name="cycle" />
						</Col>
					</Row>
				</Col>
			</Grid>
		);
	}
}

function mapStateToProps(state) {
	return {
		pattern: state.pattern,
		progress: state.progress
	};
}

function mapDispatchToProps(dispatch) {
	return {
		playprogressbar: bindActionCreators(PlayProgressBarActions, dispatch),
		patternobject: bindActionCreators(PatternActions, dispatch)
	};
}

export default connect(mapStateToProps, mapDispatchToProps)(MainView);
