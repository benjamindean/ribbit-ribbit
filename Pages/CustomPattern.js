// @flow
import React, { PureComponent } from 'react';
import { ToastAndroid, Vibration, Linking } from 'react-native';
import { Col, Row, Grid } from 'react-native-easy-grid';
import Stopwatch from 'timer-stopwatch';
import Button from '../Components/Button';
import PlayButtonFullWidth from '../Components/PlayButtonFullWidth';
import * as PlayProgressBarActions from '../Actions/PlayProgressBar';
import * as PatternActions from '../Actions/Pattern';
import * as qs from 'qs';
import styles from '../Styles/main';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

export class CustomPattern extends PureComponent {
	static navigationOptions = {
		title: 'Custom Pattern'
	};

	componentDidMount() {
		Linking.addEventListener('url', this._handleUrl);

		this.vibrateTime = new Stopwatch();
		this.pausedTime = new Stopwatch();
	}

	_handleUrl = url => {
		const queryString = url.replace(Expo.Constants.linkingUri, '');

		if (queryString) {
			const data = qs.parse(queryString);
			const pattern = data.pattern;

			if (pattern) {
				this.pattern.play(pattern);
			}
		}
	};

	_onPressIn = () => {
		this.pausedTime.stop();
		this.props.actions.pattern.update(this.pausedTime.ms);
		this.pausedTime.reset();
		this.vibrateTime.start();

		Vibration.vibrate(5000);
	};

	_onPressOut = () => {
		this.pausedTime.start();
		this.vibrateTime.stop();
		this.props.actions.pattern.update(this.vibrateTime.ms);
		this.vibrateTime.reset();

		Vibration.cancel();
	};

	_resetPattern = () => {
		this.props.actions.pattern.reset();
		this.props.actions.progressbar.reset();

		this._stopRecording();
	};

	_stopRecording() {
		this.pausedTime.stop();
		this.pausedTime.reset();
		this.vibrateTime.stop();
		this.vibrateTime.reset();
	}

	_updateProgressBar = () => {
		if (!this.props.pattern.length) {
			return;
		}

		this.props.actions.progressbar.reset();

		const percent = this.props.pattern.reduce((a, b) => a + b, 0) / 10;
		const interval = setInterval(() => {
			if (this.props.progress < 0.9) {
				this.props.actions.progressbar.update(0.1);
			} else {
				clearInterval(interval);
			}
		}, percent);
	};

	_playPattern = () => {
		this._updateProgressBar();
		this.props.actions.pattern.play(this.props.pattern);
	};

	_sharePattern = () => {
		this.props.actions.pattern.share('CustomPattern', this.props.pattern);
	};

	render() {
		return (
			<Grid style={styles.container}>
				<Col>
					<Row size={2}>
						<Col>
							<PlayButtonFullWidth 
								onPressIn={this._onPressIn} 
								onPressOut={this._onPressOut}
								progress={this.props.progress}
								pattern={this.props.pattern}
							/>
						</Col>
					</Row>
					<Row size={1} style={styles.center}>
						<Col>
							<Button onPress={this._resetPattern} name="cycle" />
						</Col>
						<Col>
							<Button onPress={this._playPattern} name="controller-play" />
						</Col>
						<Col>
							<Button onPress={this._sharePattern} name="share" />
						</Col>
					</Row>
				</Col>
			</Grid>
		);
	}
}

function mapStateToProps(state) {
	return {
		pattern: state.pattern,
		progress: state.progress
	};
}

function mapDispatchToProps(dispatch) {
	return {
		actions: {
			progressbar: bindActionCreators(PlayProgressBarActions, dispatch),
			pattern: bindActionCreators(PatternActions, dispatch)
		}
	};
}

export default connect(mapStateToProps, mapDispatchToProps)(CustomPattern);
