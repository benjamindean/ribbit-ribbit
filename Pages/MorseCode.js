// @flow
import React, { PureComponent } from 'react';
import { ToastAndroid, Vibration, View } from 'react-native';
import { Col, Row, Grid } from 'react-native-easy-grid';
import * as PlayProgressBarActions from '../Actions/PlayProgressBar';
import * as PatternActions from '../Actions/Pattern';
import styles from '../Styles/main';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

export class MorseCode extends PureComponent {
	static navigationOptions = {
		title: 'Morse Code'
	};

	render() {
		return (
			<Grid style={styles.container}>
				<Col>
				</Col>
			</Grid>
		);
	}
}

function mapStateToProps(state) {
	return {
		pattern: state.pattern,
		progress: state.progress
	};
}

function mapDispatchToProps(dispatch) {
	return {
		actions: {
			progressbar: bindActionCreators(PlayProgressBarActions, dispatch),
			pattern: bindActionCreators(PatternActions, dispatch)
		}
	};
}

export default connect(mapStateToProps, mapDispatchToProps)(MorseCode);
