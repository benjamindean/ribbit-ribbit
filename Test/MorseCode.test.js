import React from 'react';
import { MorseCode } from '../Pages/MorseCode';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
	const rendered = renderer.create(<MorseCode />).toJSON();
	expect(rendered).toMatchSnapshot();
});
