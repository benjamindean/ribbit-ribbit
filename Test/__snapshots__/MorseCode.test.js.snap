// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#3D3F4C",
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  />
</View>
`;
